numpy/random/mt19937.pyx

```diff
@@ -203,46 +147,50 @@ cdef class MT19937:
 
         Parameters
         ----------
-        seed : {None, int, array_like}, optional
+        seed : {None, int, array_like, SeedSequence}, optional
             Random seed initializing the pseudo-random number generator.
             Can be an integer in [0, 2**32-1], array of integers in
-            [0, 2**32-1] or ``None`` (the default). If `seed` is ``None``,
-            then ``MT19937`` will try to read entropy from ``/dev/urandom``
-            (or the Windows analog) if available to produce a 32-bit
-            seed. If unavailable, a 32-bit hash of the time and process
-            ID is used.
+            [0, 2**32-1], a `SeedSequence, or ``None`` (the default). If `seed`
+            is ``None``, then ``MT19937`` will try to read entropy from
+            ``/dev/urandom`` (or the Windows equivalent) if available to
+            produce a 32-bit seed. If unavailable, a 32-bit hash of the time
+            and process ID is used.
 
         Raises
         ------
         ValueError
             If seed values are out of range for the PRNG.
         """
         cdef np.ndarray obj
-        try:
-            if seed is None:
-                try:
-                    seed = random_entropy(1)
-                except RuntimeError:
-                    seed = random_entropy(1, 'fallback')
-                mt19937_seed(&self.rng_state, seed[0])
-            else:
-                if hasattr(seed, 'squeeze'):
-                    seed = seed.squeeze()
-                idx = operator.index(seed)
-                if idx > int(2**32 - 1) or idx < 0:
+        with self.lock:
+            try:
+                if seed is None:
+                    try:
+                        seed = random_entropy(1)
+                    except RuntimeError:
+                        seed = random_entropy(1, 'fallback')
+                    mt19937_seed(&self.rng_state, seed[0])
+                elif isinstance(seed, SeedSequence):
+                    val = self._seed_seq.generate_state(1, np.uint32)
```

Comment:
Please get at least 128 bits.

Reply:
is more better? Should I use RK_STATE_LEN?

Reply:
Ideally, we'd bypass `mt19937_init_by_array()` entirely by pulling `RK_STATE_LEN` `uint32` words, then [make sure the first word isn't 0](https://github.com/numpy/numpy/blob/master/numpy/random/src/mt19937/mt19937.c#L80) like in the usual routine. Then that's the state.

Reply:
That will let us replace `SeedSequence` with a `LegacyMTSeedSequence` at the Python level that replicates `mt19937_init_by_array()`. That would be a good way to implement `RandomState(seed)`.

Reply:
If following this path, should probably disallow `BitGenerators` from being input into `RandomState`, so that `RandomState(MT19937(seed))` is not allowed since it will be different from `RandomState(seed)`

Reply:
Part of NEP 19 was the requirement of being able to pass a `BitGenerator` to `RandomState` to accommodate code built for `RandomState` and not `Generator`. We want to be able to share the underlying PRNG state.

I'm not too concerned about `RandomState(MT19937(seed))` being different from `RandomState(seed)`. We should probably provide the legacy seeding somewhere more accessible, though. I happen to like `@classmethod`s for alternate constructors, myself, so we can recommend `MT19937.legacy_seeding(seed)` in the documentation. Or just `MT19937(LegacyMTSeed(seed))`.

Reply:
I would prefer not to refactor the code from `mt19937_init_by_array` into a `LegacyMTSeedSequence`.

Reply:
Okay, `@classmethod` it is, then.

Reply:
Reworked. Now `RandomState.seed` calls `MT19337.legacy_seeded`. If `RandomState` has a different BitGenerator, the call will fail. Also reworked the tests to create new Generators for each test rather than reseed.

Reply:
Hmm, [NEP 19](https://www.numpy.org/neps/nep-0019-rng-policy.html#numpy-random) says (well, _I_ said) that in such cases we should pass the provided seed through to the `BitGenerator`. Though at the time I was probably still thinking of `BitGenerator.seed()` being an API that people might use, however disrecommended it was. I think I prefer the semantics you have now, though. _If_ people are calling `np.random.seed()`, they probably wanted to reproduce the old bitstream, so making this error out when someone has threaded through another `BitGenerator` should raise the discrepancy up to them.

Reply:
If needed we can propose a change to the NEP since we now have a better picture of how this all should work